core/src/main/java/org/apache/iceberg/Partitioning.java

```diff
@@ -298,4 +331,37 @@ private static boolean compatibleTransforms(Transform<?, ?> t1, Transform<?, ?>
         || t1.equals(Transforms.alwaysNull())
         || t2.equals(Transforms.alwaysNull());
   }
+
+  // collects IDs of all partition field used across specs
+  private static Set<Integer> allFieldIds(Collection<PartitionSpec> specs) {
+    return FluentIterable.from(specs)
+        .transformAndConcat(PartitionSpec::fields)
+        .transform(PartitionField::fieldId)
+        .toSet();
+  }
+
+  // collects IDs of partition fields with non-void transforms that are present in each spec
+  private static Set<Integer> commonActiveFieldIds(Collection<PartitionSpec> specs) {
+    Set<Integer> commonActiveFieldIds = Sets.newHashSet();
+
+    int specIndex = 0;
+    for (PartitionSpec spec : specs) {
+      if (specIndex == 0) {
+        commonActiveFieldIds.addAll(activeFieldIds(spec));
+      } else {
+        commonActiveFieldIds.retainAll(activeFieldIds(spec));
```

Comment:
curious: will the same field have different field IDs across different specs? I'm reading this sentence from Iceberg specification:

> A partition field id that is used to identify a partition field and is unique within a partition spec. In v2 table metadata, it is unique across all partition specs

Reply:
There is a very very limited chance the ID is not unique in v1 tables but that should be covered by a check above. I have a test that checks we throw an exception in that case.